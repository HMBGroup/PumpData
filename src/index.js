import React from 'react'
import isPLainObject from 'is-plain-object'


export const fetchStatus = {
  failed: -1,
  initial: 0,
  loading: 1,
  ready: 2,
}

export const isReady = (status) => status === fetchStatus.ready
export const isFailed = (status) => status === fetchStatus.failed
export const isInitial = (status) => status === fetchStatus.initial
export const isLoading = (status) => status === fetchStatus.loading

const defaultOptions = {
  reducer: (state, action = {}) => action.payload,
  withClearPrevFetch: true,
}

const log = (text, mode = 'error') => (
  // eslint-disable-next-line prefer-template
  console[mode]('[react-local-fetch]: ' + text)
)

export const withLocalFetch = (requestName, options) => (
  (BaseComponent) => (
    class WithLocalFetch extends React.Component {
      constructor(props) {
        super(props)
        this.options = {
          ...defaultOptions,
          ...(
            typeof options === 'function'
              ? options(props, this.getProps) : options
          ),
        }
        this.requestName = requestName
        this.state = {
          data: this.options.reducer(undefined, {}),
          status: fetchStatus.initial,
          error: null,
        }
        this.isMountedMain = false
        this.fetchId = 1
      }

      componentDidMount() {
        this.isMountedMain = true
      }

      componentWillUnmount() {
        this.isMountedMain = false
      }

      getProps = () => this.props

      fetch = async ({ type, ...args } = {}) => {
        this.fetchId += 1
        const localFetchId = this.fetchId

        try {
          this.setState({ status: fetchStatus.loading })
          const result = await this.options.action(args)

          if (!this.isMountedMain || !this.isSameRequest(localFetchId)) {
            return undefined
          }

          this.setState((prev) => ({
            status: fetchStatus.ready,
            data: this.options
              .reducer(prev.data, { type, payload: result }),
          }))

          return result
        }
        catch (error) {
          if (this.isMountedMain && this.isSameRequest(localFetchId)) {
            this.setState({ status: fetchStatus.failed, error })
          }

          return undefined
        }
      }

      isSameRequest = (fetchid) => (
        this.options.withClearPrevFetch
          ? this.fetchId === fetchid : true
      )

      dispatch = (action) => {
        if (!isPLainObject(action)) {
          log('action must be plain object')
          return
        }

        const computedData = this.options
          .reducer(this.state.data, action)

        this.setState({ data: computedData })
      }

      render() {
        return React
          .createElement(BaseComponent, {
            ...this.props,
            [this.requestName]: {
              ...this.state,
              fetch: this.fetch,
              dispatch: this.dispatch,
            },
          })
      }
    }
  )
)
