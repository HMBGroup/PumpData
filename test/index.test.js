import React from 'react'
import Enzyme, { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import { withLocalFetch, fetchStatus } from '../src/index'


Enzyme.configure({ adapter: new Adapter() })

let viewStub = null

beforeEach(() => {
  viewStub = ({ news, users }) => (
    <div>
      <button
        onClick={() => news.fetch({ id: 1 })}
        testid="fetchNews"
      >
        fetch news
      </button>
      <button
        onClick={() => users.fetch({ id: 2 })}
        testid="fetchUsers"
      >
        fetch users
      </button>
    </div>
  )
})

test('Should hoc render without error', () => {
  const MyComponent = withLocalFetch()(() => <p>hello</p>)
  const wrapper = renderer.create(<MyComponent />)

  expect(wrapper).toMatchSnapshot()
})

test('Should request function was called', async () => {
  const mockRequest = jest.fn()
  const requests = { news: mockRequest, users: mockRequest }
  const enhance = withLocalFetch(requests)
  const TestComponent = enhance(viewStub)
  const wrapper = mount(<TestComponent />)
  const newsBtn = wrapper.find('[testid="fetchNews"]')
  const usersBtn = wrapper.find('[testid="fetchUsers"]')

  newsBtn.simulate('click')
  usersBtn.simulate('click')

  expect(requests.news.mock.calls.length).toBe(2)
})

test('Mapping fetchStatus and fetch result status', async () => {
  const mockRequest = jest.fn()
  const mockRequestWithThrow = () => {
    throw new Error('some error message')
  }
  const requests = { news: mockRequest, users: mockRequestWithThrow }
  const enhance = withLocalFetch(requests)
  const TestComponent = enhance(viewStub)
  const wrapper = mount(<TestComponent />)
  const newsBtn = wrapper.find('[testid="fetchNews"]')
  const usersBtn = wrapper.find('[testid="fetchUsers"]')

  await newsBtn.simulate('click')
  await usersBtn.simulate('click')

  const { state } = wrapper.instance()

  expect(state.requests.news.status).toBe(fetchStatus.ready)
  expect(state.requests.users.status).toBe(fetchStatus.fail)
})
