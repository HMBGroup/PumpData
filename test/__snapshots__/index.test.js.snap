// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should hoc render without error 1`] = `
<p>
  hello
</p>
`;
